contrib/lkt/language/parser.py

```diff
@@ -1862,16 +1901,187 @@ def check_correctness():
+                    ),
+
+                    # If they're the same, iterate on the actuals of both
+                    # types, and call ``infer_actuals_impl`` recursively,
+                    # searching for generic formals on the gen type side.
+                    inst.actuals.map(
+                        lambda i, actual: Entity.infer_actuals_impl(
+                            generic_decl,
+                            actual.as_bare_entity,
+                            etype.actuals.at(i).as_bare_entity
+                        )
+                    ).then(
+                        # If the result of any subresult is an error, return it
+                        lambda subresults:
+                        subresults.find(lambda sr: Not(sr.error.is_null)).then(
+                            lambda sr: sr,
+                            # Else, concatenate the subresults and return that
+                            default_val=InferInstantiation.new(
+                                result=subresults.mapcat(lambda sr: sr.result)
+                            )
+                        ),
+                    ),
+                ),
+
+                # If the expected_type is not a generic, error
+                default_val=InferInstantiation.new(
+                    error=SemanticResult.new(has_error=True)
+                ),
+            ),
+
+            # General case: the gen type is neither a generic formal nor an
+            # instantiated generic type: we have nothing to do, return an empty
+            # result.
+            lambda _: No(InferInstantiation)
         )
 
-    @langkit_property()
+    @langkit_property(return_type=T.InferInstantiation)
+    def infer_actuals(
+        generic_decl=T.GenericDecl.entity,
+        expected_type=T.TypeDecl.entity
+    ):
+        """
+        Infer the actuals for ``generic_decl``, given an ``expected_type`` for
+        the call of the declaration. Return an error result if incorrect.
+        """
+        unsorted_actuals = Var(Entity.infer_actuals_impl(
+            generic_decl,
+            generic_decl.decl.called_decl_type, expected_type
+        ))
+
+        return Cond(
+            # Inference returned an error without message -> emit a general
+            # unification error.
+            Not(unsorted_actuals.error.is_null)
+            & unsorted_actuals.error.error_message.is_null,
+            InferInstantiation.new(error=Self.error(
+                S("Cannot instantiate callee: cannot unify `")
+                .concat(generic_decl.decl.called_decl_type.full_name)
+                .concat(S("` with `")).concat(expected_type.full_name)
+                .concat(S("`"))
+            )),
+
+            # Inference returned an error with message -> propagate it
+            Not(unsorted_actuals.error.is_null),
+            unsorted_actuals,
+
+            # We don't have all the actuals needed to infer -> return an error
+            unsorted_actuals.result.length
```

Comment:
It seems to me like this check is too soft. Consider for example:
```
def mkPair[T](x: T) -> Pair[T, T]

def foo(): Pair[Int, Bool] = mkPair(1)
```
Am I right that `infer_actuals_impl` would return `[T -> Int, T -> Bool]` here? So, this check would pass and further down in `check_infer_actuals` we will take the first actual for `T`, so `Int`, meaning the whole thing will typecheck "successfully".

So shouldn't we also check that the mappings returned by `infer_actuals_impl` contain no incoherent binding?

Reply:
That's correct, there is a missing check, good catch

I'll implement it in a further PR, since globally WIP anyway